core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
+    options = options or (type(content) == "table" and content.attr) or {}
+  })
+end
+--- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushText(text, line, column, file)
+  return self:_pushFrame({
+    text = text,
+    file = file,
+    line = line,
+    column = column
+  })
+end
+--- Internal: Push complete frame
+function SILE.currentCommandStack:_pushFrame(frame)
+  frame.toString = self._frameToString
+  frame.typesetter = SILE.typesetter
+  --frame.pushTrace = debug.traceback(nil, 3)
+
+  -- Push the frame
+  if SU.debugging("commandStack") then
+    print(string.rep(" ", #self) .. "PUSH(" .. frame:toString(false, true) .. ")")
+  end
+  self[#self + 1] = frame
+  self.lastCommand = nil
+  self._lastPushId = self._lastPushId + 1
+  frame.pushId = self._lastPushId
+  return self._lastPushId
+end
+--- Pop previously pushed command from the stack.
+--- Return value of `push` function must be provided as argument to check for balanced usage.
+function SILE.currentCommandStack:pop(pushId)
+  if type(pushId) ~= "number" then
+    SU.error("SILE.currentCommandStack:pop's argument must be the result value of the corresponding push", true)
+  end
+
+  -- First verify that push/pop is balanced
+  local popped = self[#self]
+  if popped.pushId ~= pushId then
+    local message = "Unbalanced content push/pop"
+    local debug = SILE.detailedErrors or SU.debugging("commandStack")
+    if debug then
+      message = message .. ". Expected " .. popped.pushId .. " - (" .. popped:toString() .. "), got " .. pushId
+    end
+    SU.warn(message, debug)
+  else
+    -- Correctly balanced: pop the frame
+    self.lastCommand = popped
+    self[#self] = nil
+    if SU.debugging("commandStack") then
+      print(string.rep(" ", #self) .. "POP(" .. popped:toString(false, true) .. ")")
+    end
+  end
+end
+--- Returns a stack of commands, which can be used to show information about location in document.
+--- Most relevant object is in the last index.
+--- Second return: recently popped value which could provide extra location information
+function SILE.currentCommandStack:locationStack(currentTypesetterOnly)
+  local result = {}
+
+  for i = 1, #self do
+    if not currentTypesetterOnly or (self[i].typesetter == SILE.typesetter) then
+      result[#result + 1] = self[i]
+    end
+  end
+
+  local extra = self.lastCommand
+  if (not extra) -- skip other checks if extra is nil
+          or extra == result[#result] -- no extra if it already is on the stack
+          or (currentTypesetterOnly and (extra.typesetter ~= SILE.typesetter)) -- no extra if wrong typesetter
+  then
+    extra = nil
+  end
+
+  return result, extra
+end
+
+--- Internal: Format top of the stack from locationStack(), may return nil
+function SILE.currentCommandStack:_locationTraceEntry(stack, after)
```

Comment:
after what?

Reply:
This is a second return value of `locationStack()` (bad name, I know). 
It is a frame which was popped recently and its description will be appended to the result. It is called "after", because the SILE's position in the document is somewhere *after* that command.

It is hard to think up names for these things.